### PR TITLE
[mffan] Initial contribution

### PR DESCRIPTION
**Pull request by mark-brooks-180**
I've developed a new binding for "Modern Forms" and "WAC Lighting" ceiling fans.  The binding implements communications , and control of these WIFI connected, smart fans.

### CODE REVIEW

File: bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        }
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        try {
+            this.executor.execute(new CommandHandlerRunnable(channelUID, command));
+        } catch (@SuppressWarnings("unused") Exception e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java:113]: see earlier comment

File: bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                    String.format("Could not control device at IP address %s", this.config.getIpAddress()));
+        }
+    }
+
+    private void handleCommand(Command command) {
+        handleCommand(new ChannelUID("Null:Null:Null:Null"), command);
+    }
+
+    private final class CommandHandlerRunnable implements Runnable {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java:122]: Might be valid, but this is not a common pattern i see in most bindings, can you explain a bit on why it is moddled this way and not like many other bindings handling the commands inside the handdler class?

File: bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/RestApiException.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * The {@link RestApiException} is an exception thrown from the REST API.
+ *
+ * @author Mark Brooks - Initial contribution
+ */
+@NonNullByDefault
+public class RestApiException extends RuntimeException {
+    private static final long serialVersionUID = 1L;
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/RestApiException.java:24]: Would be nice if you can create/generate  a serialVersionUID other than 1L

File: bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/ShadowBufferDto.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
+/**
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/ShadowBufferDto.java:1]: I would suggest to move the API to a different folder and package. 
bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/api/
Files `FanRestApi.java`, `RestApiException.java` and `ShadowBufferDto.java`

File: bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/addon/addon.xml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<addon:addon id="mffan" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:addon="https://openhab.org/schemas/addon/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/addon/v1.0.0 https://openhab.org/schemas/addon-1.0.0.xsd">
+
+	<type>binding</type>
+	<name>Modern Forms Fan Binding</name>
+	<description>This is the binding for "Modern Forms", and "WAC Lighting" smart ceiling fans.</description>
+
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/addon/addon.xml:9]: Can you add connection tag and maybe country and discovery ? (last two might be more difficult)

File: bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/i18n/mffan.properties [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# FIXME: please add all English translations to this file so the texts can be translated using Crowdin
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/i18n/mffan.properties:1]: FIXME :-)

File: bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/thing/mffan.xml [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="mffan">
+
+		<label>Modern Forms Fan</label>
+		<description>Modern Forms and WAC Lighting Smart Ceiling Fans</description>
+
+		<channels>
+			<channel id="fanOn" typeId="fanOn"/>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/thing/mffan.xml:13]: See earlier comment about channel id's that should be lower-case-hyphen

File: bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/thing/mffan.xml [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+			<parameter name="pollingPeriod" type="integer" unit="s" min="1">
+				<label>Refresh Interval</label>
+				<description>Interval the device is polled in seconds.</description>
+				<default>120</default>
+				<advanced>true</advanced>
+			</parameter>
+		</config-description>
+	</thing-type>
+
+	<channel-type id="fanOn">
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/thing/mffan.xml:37]: lower-case-hyphen naming convention also applies to channel type id's

File: bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/thing/mffan.xml [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+			</options>
+		</state>
+	</channel-type>
+	<channel-type id="lightOn">
+		<item-type>Switch</item-type>
+		<label>Light</label>
+		<description>Light on/off</description>
+	</channel-type>
+	<channel-type id="lightIntensity">
+		<item-type>Dimmer</item-type>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/thing/mffan.xml:91]:
```suggestion
		<item-type>Number:Dimensionless</item-type>
```

File: bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/thing/mffan.xml [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+		<label>Light Intensity</label>
+		<description>Light intensity</description>
+		<state readOnly="false" min="1" max="100" step="1"/>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/resources/OH-INF/thing/mffan.xml:94]:
```suggestion
		<state readOnly="false" min="1" max="100" step="1"  pattern="%.0f %unit%" />
```

File: bundles/pom.xml [hunk 11]
```diff
@@ -1,690 +1,2706 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (lsiepel) [inline bundles/pom.xml:1]: Something weird happend to this file, please revert / only change the line to add this binding.

File: bundles/org.openhab.binding.mffan/README.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
+# MfFan Binding
+
+This binding is used to enable communications between openHAB and "Modern Forms" or "WAC Lighting" WIFI connected, smart, ceiling fans.  
+
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/README.md:4]:
```suggestion
```

File: bundles/org.openhab.binding.mffan/README.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+## Supported Things
+
+Supported things include: "Modern Forms" and "WAC Lighting" smart ceiling fans.  The binding has been tested with both.
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/README.md:8]: This are the supported devices, not the supported `Things`. Maybe add these devices as tested or somehting and use the Thing Id listed in the thing-type.xml

File: bundles/org.openhab.binding.mffan/README.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+## Discovery
+
+Auto discovery is not supported at this time.  
+To bind with a smart fan either the IP address of the fan, or the local host name of the fan must be provided.
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/README.md:13]:
```suggestion
```
Those details belong to the configuration section

File: bundles/org.openhab.binding.mffan/README.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+The binding has been tested using an IPV4 local network address, and a host name served by a local DNS.
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/README.md:14]:
```suggestion
```
Not sure if this should be mentioned at all as the configuration explains hsotname/ipadres (maybe add IPV4 to that descritpion)

File: bundles/org.openhab.binding.mffan/README.md [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+## Binding Configuration
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/README.md:16]:
```suggestion
## Thing Configuration
```

File: bundles/org.openhab.binding.mffan/README.md [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+| Name            | Type    | Description                           | Default | Required | Advanced |
+|-----------------|---------|---------------------------------------|---------|----------|----------|
+| hostname        | text    | Hostname or IP address of the device  | N/A     | yes      | no       |
+| refreshInterval | integer | Interval the device is polled in sec. | 120     | no       | yes      |
+
+## Channels
+
+| Channel        | Type   | Read/Write | Description                         |
+|----------------|--------|------------|-------------------------------------|
+| fanOn          | Switch | RW         | Channel that turns the fan on/off.  |
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/README.md:28]: See other comment about lower-case-hyphen convention

**Review by lsiepel — CHANGES_REQUESTED** [review]
Thanks for your contribution. I'm not familiair with this device, but it looks familiair to the [bigassfan ](https://www.openhab.org/addons/bindings/bigassfan/) binding. Maybe that is something you can use for inspiration.

Anyway, besides the comments, you should add yourself to the codeowners file and add this binding to the bom/openhab-addons/pom.xml . I prefer that you respond (`done` is fine) to the comments and let me mark them as resolved. 
Let me know if you need anything.

**Comment by mark-brooks-180** [comment]
Thanks for your prompt reply.  I'll make the suggested changes.  I'll let
you know when I'm done.

Thank,
Mark B.

**Comment by mark-brooks-180** [comment]
------------------------------


lsiepel,

I believe I have all of the changes you recommended completed.  The
corruption of file In bundles/pom.xml
<https://github.com/openhab/openhab-addons/pull/16786#discussion_r1610444528>
was
resolved by resynchronizing my fork with the upstream and then recommitting
my changes.  The two issues listed below require some explanation:

  - I'm using a single thread executor to serialize command handler
execution.

  - The actual commands are executed by runnables submitted to the
executor.  This ensures the commands are executed in the order they are
received, and potential delays resulting from calls to the REST API
functions don't impact the performance of calls to handleCommand(ChannelUID
channelUID, Command command).

  - In a similar manner the polling thread calls a private version of
handleCommand to submit a REFRESH command.  This ensures commands are
serialized by the executor, and avoids any thread contention when updating
the state of the Thing.

Thanks,

Mark B.

In
bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java
<https://github.com/openhab/openhab-addons/pull/16786#discussion_r1610420282>
:

> +            // this.pollingJob = this.scheduler.scheduleWithFixedDelay(() -> update(this.api.getShadowBuffer()), 0,
+            // pollingPeriod, TimeUnit.SECONDS);

⬇️ Suggested change

-            // this.pollingJob =
this.scheduler.scheduleWithFixedDelay(() ->
update(this.api.getShadowBuffer()), 0,
-            // pollingPeriod, TimeUnit.SECONDS);



In
bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java
<https://github.com/openhab/openhab-addons/pull/16786#discussion_r1610428827>
:

> +    @SuppressWarnings("null")
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        try {
+            this.executor.execute(new
CommandHandlerRunnable(channelUID, command));
+        } catch ***@***.***("unused") Exception e) {
+            updateStatus(ThingStatus.OFFLINE,
ThingStatusDetail.COMMUNICATION_ERROR,
+                    String.format("Could not control device at IP
address %s", this.config.getIpAddress()));
+        }
+    }
+
+    private void handleCommand(Command command) {
+        handleCommand(new ChannelUID("Null:Null:Null:Null"), command);
+    }
+
+    private final class CommandHandlerRunnable implements Runnable {

Might be valid, but this is not a common pattern i see in most bindings,
can you explain a bit on why it is moddled this way and not like many other
bindings handling the commands inside the handdler class?

**Comment by lsiepel** [comment]
I think something went wrong as the pr is closed due to no files. It should not be needed to do a force push. Can you fix this?

**Comment by mark-brooks-180** [comment]
Isiepel,

I believe I fixed the issue.  The pull request has been reopened.

Thanks,
Mark B.

**Review by lsiepel — CHANGES_REQUESTED** [review]
Some, but not all previous comments are addressed. I also see some comments being lost after the changes. Ping me when your are ready.

**Comment by mark-brooks-180** [comment]
Isiepel,

I believe I've made all the recommended changes and pushed them to main.  I
tested the binding.  Everything looks good on my end.

Thanks,
Mark B.

File: bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java [hunk 18]
```diff
@@ -72,22 +73,15 @@ public void initialize() {
...
-        if (validateConfig(this.config)) {
-            try {
-                this.api = new FanRestApi(this.config.getIpAddress().trim(), this.httpClientFactory);
-                // handleCommand(RefreshType.REFRESH);
-            } catch (Exception e) {
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.getMessage());
-            }
-
+        if (MfFanConfiguration.validateConfig(this.config)) {
+            this.api = new FanRestApi(this.config.getIpAddress().trim(), this.httpClientFactory);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java:78]: Minor: The trim() should be part of the getIpAddress() as that method is responsible for returning a proper IpAddress

File: bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+        this.httpClientFactory = httpClientFactory;
+        this.executor = Executors.newSingleThreadExecutor();
+    }
+
+    @Override
+    public void initialize() {
+        this.logger.debug("Initializing MfFan handler '{}'", getThing().getUID());
+        updateStatus(ThingStatus.UNKNOWN);
+        this.config = getConfigAs(MfFanConfiguration.class);
+        if (MfFanConfiguration.validateConfig(this.config)) {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java:77]:
```suggestion
        if (!MfFanConfiguration.validateConfig(this.config)) {
        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Invalid configuration detected.");
        return;
        }
```
Better to abort immediate. (last 3 lines of this method can be removed.

You could also consider to use a custom i18n key.
Reference: https://www.openhab.org/docs/developer/utils/i18n.html#using-custom-keys (last example)

File: bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/MfFanConfiguration.java [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+    public Integer getPollingPeriod() {
+        return this.pollingPeriod;
+    }
+
+    public void setPollingPeriod(Integer pollingPeriod) {
+        this.pollingPeriod = pollingPeriod;
+    }
+
+    public static boolean validateConfig(@Nullable MfFanConfiguration config) {
+        if ((config == null) || config.getIpAddress().isBlank()) {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/MfFanConfiguration.java:51]:
```suggestion
        if (config == null || config.getIpAddress().isBlank()) {
```

File: bundles/org.openhab.binding.mffan/README.md (hunk 12 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.mffan/README.md:4]: Not fixed yet

**Comment by mark-brooks-180** [comment]
Isiepel,

I believe I have made the recommended changes.  Furthermore, I eliminated
the use of the single thread executor from the MfFanHandler class.  Your
email had a bullet item that said

   - fix the build (see bottom of this page)

I didn't find anything at the bottom of the page indicating the build error
you are experiencing.

Thanks,
Mark B.

File: bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    @Override
+    public void dispose() {
+        this.logger.debug("Disposing MF fan handler '{}'", getThing().getUID());
+        ScheduledFuture<?> job = this.pollingJob;
+        if (job != null) {
+            job.cancel(true);
+        }
+        this.pollingJob = null;
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mffan/src/main/java/org/openhab/binding/mffan/internal/handler/MfFanHandler.java:90]:
```suggestion
        this.pollingJob = null;
        }
```

**Review by lsiepel — CHANGES_REQUESTED** [review]
Many comments are fixed, thanks. Left one new comment and besides that
- DCO is not signed (mandatory)
- build failed (run spotless)
- Unalbe to check now, but fix any SAT errors
- Unable to check now but fix any compile warnings

**Comment by mark-brooks-180** [comment]
Isiepel,

I've pushed the most recent set of changes.  Let me know if there is
anything else I need to do.

Thanks,
Mark B.

**Comment by lsiepel** [comment]
You could fix the errors below. (ALS listed in my previous comment)

**Comment by mark-brooks-180** [comment]
Fixed.

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM.

Only remaining issue is the DCO, if you click details next to the DCO you can see the related commits and how to fix those.

**Comment by lsiepel** [comment]
> Thanks, LGTM.
> 
> Only remaining issue is the DCO, if you click details next to the DCO you can see the related commits and how to fix those.

@mark-brooks-180 you need any help with the DCO?

**Comment by mark-brooks-180** [comment]
Isiepel,

I believe the DCO issue is resolved.  All the commits are signed off.

M

**Comment by lsiepel** [comment]
Now, you could add your binding's logo to the openHAB website. See https://next.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website-and-the-ui

**Comment by mark-brooks-180** [comment]
Isiepel,

Thanks for your help.

M

**Pull request merged**